v2/internal/crdmanagement/manager.go

```diff
@@ -154,70 +155,142 @@ func (m *Manager) FindNonMatchingCRDs(
+		if !ok {
+			return nil, errors.Errorf("Couldn't find goal CRD %q. This is unexpected!", name)
+		}
+
+		result.DiffResult = SpecDifferent
 	}
-	for name, crd := range goalCRDsWithDifferentSpec {
-		crdsToApply[name] = crd
+	for name := range goalCRDsWithDifferentVersion {
+		result, ok := resultMap[name]
+		if !ok {
+			return nil, errors.Errorf("Couldn't find goal CRD %q. This is unexpected!", name)
+		}
+
+		result.DiffResult = VersionDifferent
 	}
 
-	if len(crdsToApply) > 0 {
-		m.logger.V(Status).Info("Will apply CRDs", "count", len(crdsToApply))
+	// Collapse result to a slice
+	results := make([]*CRDInstallationInstruction, 0, len(resultMap))
+	for _, result := range resultMap {
+		results = append(results, result)
+	}
 
-		i := 0
-		for _, crd := range crdsToApply {
-			crd := crd
+	return results, nil
+}
 
-			i += 1
-			_, isDifferentVersion := goalCRDsWithDifferentVersion[crd.Name]
-			_, isDifferentSpec := goalCRDsWithDifferentSpec[crd.Name]
+func (m *Manager) ApplyCRDs(
+	ctx context.Context,
+	instructions []*CRDInstallationInstruction,
+) error {
+	var instructionsToApply []*CRDInstallationInstruction
+
+	for _, item := range instructions {
+		apply, reason := item.ShouldApply()
+		if apply {
+			instructionsToApply = append(instructionsToApply, item)
 			m.logger.V(Verbose).Info(
-				"Found CRD in need of update",
-				"Progress", fmt.Sprintf("%d/%d", i, len(crdsToApply)),
-				"CRD", crd.Name,
-				"SpecDifferent", isDifferentSpec,
-				"VersionDifferent", isDifferentVersion)
-
-			toApply := &apiextensions.CustomResourceDefinition{
-				ObjectMeta: metav1.ObjectMeta{
-					Name: crd.Name,
-				},
-			}
-			result, err := controllerutil.CreateOrUpdate(ctx, m.kubeClient, toApply, func() error {
-				resourceVersion := toApply.ResourceVersion
-				*toApply = crd
-				toApply.ResourceVersion = resourceVersion
+				"Will update CRD",
+				"crd", item.CRD.Name,
+				"diffResult", item.DiffResult,
+				"filterReason", item.FilterReason,
+				"reason", reason)
+		} else {
+			m.logger.V(Verbose).Info(
+				"Will NOT update CRD",
+				"crd", item.CRD.Name,
+				"reason", reason)
+		}
+	}
 
-				return nil
-			})
-			if err != nil {
-				return errors.Wrapf(err, "failed to apply CRD %s", crd.Name)
-			}
+	if len(instructionsToApply) == 0 {
+		m.logger.V(Status).Info("Successfully reconciled CRDs because there were no CRDs to update.")
+		return nil
+	}
+
+	m.logger.V(Status).Info("Will apply CRDs", "count", len(instructionsToApply))
 
-			m.logger.V(Debug).Info("Successfully applied CRD", "name", crd.Name, "result", result)
+	i := 0
+	for _, instruction := range instructionsToApply {
+		instruction := instruction
+
+		i += 1
+		toApply := &apiextensions.CustomResourceDefinition{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: instruction.CRD.Name,
+			},
+		}
+		m.logger.V(Verbose).Info(
+			"Applying CRD",
+			"Progress", fmt.Sprintf("%d/%d", i, len(instructionsToApply)),
+			"crd", instruction.CRD.Name)
```

Comment:
Names should both Capitalized or both lowercase.

```suggestion
			"progress", fmt.Sprintf("%d/%d", i, len(instructionsToApply)),
			"crd", instruction.CRD.Name)
```

Reply:
I don't know how the pedant in me missed this, usually this sort of thing is super obvious (if obviously pretty minor). Thanks for the catch.